formats/json-tests/commonTest/src/kotlinx/serialization/json/JsonExponentTest.kt

```diff
@@ -0,0 +1,79 @@
+package kotlinx.serialization.json
+
+import kotlinx.serialization.Serializable
+import kotlinx.serialization.test.*
+import kotlin.test.Test
+import kotlin.test.assertEquals
+
+class JsonExponentTest : JsonTestBase() {
+    @Serializable
+    data class SomeData(val count: Long)
+    @Serializable
+    data class SomeDataDouble(val count: Double)
+
+    @Test
+    fun testExponentDecodingPositive() = parametrizedTest {
+        val decoded =  Json.decodeFromString<SomeData>("""{ "count": 23e11 }""", it)
+        assertEquals(2300000000000, decoded.count)
+    }
+
+    @Test
+    fun testExponentDecodingNegative() = parametrizedTest {
+        val decoded = Json.decodeFromString<SomeData>("""{ "count": -10E1 }""", it)
+        assertEquals(-100, decoded.count)
+    }
+
+    @Test
+    fun testExponentDecodingTruncatedDecimal() = parametrizedTest {
+        val decoded = Json.decodeFromString<SomeData>("""{ "count": -1E-1 }""", it)
```

Comment:
I've just realized that it is incorrect to truncate such values. That's why the problem with ` 4.9E-324` happens in the first place - it is too small. We should not allow such values to be parsed to `Long`, or any value with non-zero fractional part after applying exponent. I think the easiest way to check it would be with `doubleAccumulator.floor() == doubleAccumulator`, as in here: https://github.com/Kotlin/kotlinx.serialization/blob/e2092a41693c79686e787ccd975579aabb0b2412/formats/json/jsMain/src/kotlinx/serialization/json/internal/DynamicDecoders.kt#L154